server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java

```diff
@@ -3925,11 +3924,16 @@ public void testScheduledRefresh() throws Exception {
         logger.info("--> ensure search idle");
         assertTrue(primary.isSearchIdle());
         assertTrue(primary.searchIdleTime() >= TimeValue.ZERO.millis());
+        long periodicFlushesBefore = primary.flushStats().getPeriodic();
         primary.flushOnIdle(0);
+        assertBusy(() -> assertThat(primary.flushStats().getPeriodic(), greaterThan(periodicFlushesBefore)));
+
+        long externalRefreshesBefore = primary.refreshStats().getExternalTotal();
         logger.info("--> scheduledRefresh(future5)");
-        PlainActionFuture<Boolean> future5 = new PlainActionFuture<>();
-        primary.scheduledRefresh(future5);
-        assertTrue(future5.actionGet()); // make sure we refresh once the shard is inactive
+        primary.scheduledRefresh(ActionListener.noop());
+        // We can't check whether scheduledRefresh returns true because it races with a potential
+        // refresh triggered by the flush. We just check that one the refreshes ultimately wins.
```

Comment:
@kingherc I'm not sure what we can investigate. There are always a probability of a refresh whether from a flush or from some other process. We can't deterministically check whether an individual refresh request will succeed. We can't even realistically control when the flush will be executed, since it's also asynchronous.

The test is suppose to check that a refresh goes through from `IndexShard#scheduledRefresh` to `engine.maybeRefresh` when the shard is no longer idle. Since don't have enough isolation to test that, we have to rely on eihter re-trying refreshes until one succeeds or checking refresh stats.

Reply:
If we think that the problem is the concurrent flush (I don't think that's the issue, since the flush logs _after_ the test fails) we should either try to reproduce it with sleeps or maybe inject a deterministic thread pool that allows us to control the execution order. Otherwise we risk not fixing the real issue for this test.

Reply:
@fcofdez The original test was never suppose to test that, it just tested that the refresh is actually gets triggered when the shard isn't idle. Here's original pull request from 2017 https://github.com/elastic/elasticsearch/pull/27500.

I still think it doesn't matter whether another refresh gets triggered by a flush or another background refresh, because we can't guarantee that any individual refresh call will return `true`.

So, the current change tries to mimic the original scenario for this PR (a sync flush followed by a refresh and check that some refresh has completed).

Reply:
Should we sync on this in a short call?